source/common/formatter/http_specific_formatter.cc

```diff
@@ -29,16 +29,22 @@ HttpFormatterContext::HttpFormatterContext(const Http::RequestHeaderMap* request
       log_type_(log_type) {}
 
 const Http::RequestHeaderMap& HttpFormatterContext::requestHeaders() const {
-  return request_headers_ != nullptr ? *request_headers_
-                                     : *Http::StaticEmptyHeaders::get().request_headers;
+  if (request_headers_ == nullptr) {
+    request_headers_ = Http::StaticEmptyHeaders::get().request_headers.get();
+  }
+  return *request_headers_;
```

Comment:
Are these change for any particular reason?

Reply:
The requestHeaders/responseHeaders methods maybe called multiple times in same formatting, and this change could ensure the `StaticEmptyHeaders::get()` only be called once.

These empty headers are constructed on first use. So, the `StaticEmptyHeaders::get()` require at least one time atomic bool checking.
This change want to reduce these atomic operations.

This is pretty simple change and from the benchmark results, this change at least not make things worse.

(I actually have no insistence on this point. If you think this is unnecessary, I can revert it or add a new benchmark case to ensure it's effect.)

Reply:
Ah, I see. I'm not an enormous fan of mutable members if we can avoid them. I'd be inclined to revert this part of the change and write a benchmark to see if it makes a difference, and then reassess. But up to you.